### PR TITLE
[python-package] Some mypy fixes

### PR DESCRIPTION
**Pull request by AlbertoEAF**
Related to https://github.com/microsoft/LightGBM/issues/3867.

Fixes some mypy false-positives.
Mypy is not smart enough to detect some things like alternative definitions to imports. Some type ignores are required.

### CODE REVIEW

File: python-package/lightgbm/callback.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ def __init__(self, best_iteration, best_score):
 
 # Callback environment used by callbacks
 CallbackEnv = collections.namedtuple(
-    "LightGBMCallbackEnv",
+    "CallbackEnv",
```
Comment (jameslamb) [inline python-package/lightgbm/callback.py:29]: what mypy warning does this address? I'm confused by this change.

File: python-package/lightgbm/sklearn.py [hunk 2]
```diff
@@ -791,7 +791,7 @@ def fit(self, X, y,
                     categorical_feature=categorical_feature, callbacks=callbacks, init_model=init_model)
         return self
 
-    _base_doc = LGBMModel.fit.__doc__
+    _base_doc = LGBMModel.fit.__doc__ or ""
     _base_doc = (_base_doc[:_base_doc.find('group :')]
```
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:795]: can you please revert these `_base_doc` fixes? Another contributor is already helping with them on #3891

File: python-package/setup.py [hunk 3]
```diff
@@ -322,7 +322,7 @@ def run(self):
     if os.path.isfile(os.path.join(CURRENT_DIR, os.path.pardir, 'VERSION.txt')):
         copy_file(os.path.join(CURRENT_DIR, os.path.pardir, 'VERSION.txt'),
                   os.path.join(CURRENT_DIR, 'lightgbm', 'VERSION.txt'),
-                  verbose=0)
+                  verbose=0) # type: ignore
```
Comment (jameslamb) [inline python-package/setup.py:325]: can you please revert this too? #3891 is already fixing this

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks! Please see my comments. Can you also please share the specific mypy warnings your PR addresses? That will help us to understand the changes.

File: python-package/lightgbm/callback.py (hunk 1 above)
Comment (AlbertoEAF) [reply python-package/lightgbm/callback.py:29]: The declared type `LightGBMCallbackEnv` is not the same name as the variable that holds that definition. That is unnecessary and confusing. Mypy complains about that.

File: python-package/lightgbm/sklearn.py (hunk 2 above)
Comment (AlbertoEAF) [reply python-package/lightgbm/sklearn.py]: done

File: python-package/lightgbm/callback.py (hunk 1 above)
Comment (jameslamb) [reply python-package/lightgbm/callback.py:29]: I was asking for the specific warning. I ran `mypy --ignore-missing imports python-package/`, and looks like it is this

> python-package/lightgbm/callback.py:28: error: First argument to namedtuple() should be 'CallbackEnv', not 'LightGBMCallbackEnv'

File: python-package/lightgbm/compat.py [hunk 4]
```diff
@@ -11,16 +11,12 @@
 except ImportError:
     PANDAS_INSTALLED = False
 
-    class pd_Series:
+    class pd_Series: # type: ignore
```
Comment (jameslamb) [inline python-package/lightgbm/compat.py:14]:
```suggestion
    class pd_Series:  # type: ignore
```

please add two spaces before each of these comments, to match our linting rules

https://github.com/microsoft/LightGBM/pull/3916/checks?check_run_id=1894201789

> ./python-package/lightgbm/compat.py:14:21: E261 at least two spaces before inline comment
./python-package/lightgbm/compat.py:17:24: E261 at least two spaces before inline comment
./python-package/lightgbm/compat.py:48:24: E261 at least two spaces before inline comment
./python-package/lightgbm/compat.py:122:22: E261 at least two spaces before inline comment
./python-package/lightgbm/compat.py:125:26: E261 at least two spaces before inline comment
./python-package/lightgbm/compat.py:128:23: E261 at least two spaces before inline comment

File: python-package/lightgbm/compat.py [hunk 5]
```diff
@@ -11,16 +11,12 @@
...
         """Dummy class for pandas.Series."""
 
-        pass
```
Comment (jameslamb) [inline python-package/lightgbm/compat.py:17]: can you please put these `pass` statements back? I understand that they aren't strictly necessary and that some linters like `pylint` will tell you to remove them, but we have a preference for using `pass` statements in cases like this to make it clear more explicit that the classes don't do anything by themselves.

`mypy` does not care about these, as far as I know, so I don't think the change has an impact on `mypy`'s output

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks! I left a few additional suggestions.

**Comment by jameslamb** [comment]
@AlbertoEAF are you still interested in helping out on this?

**Comment by AlbertoEAF** [comment]
Thanks for the ping James, done and rebased ;)

**Review by jameslamb — CHANGES_REQUESTED** [review]
Looks good, thanks! Can you please just revert the extra lines you've added?

That will resolve these linting errors.

```text
./python-package/lightgbm/compat.py:20:5: E303 too many blank lines (2)
./python-package/lightgbm/compat.py:26:5: E303 too many blank lines (2)
./python-package/lightgbm/compat.py:134:5: E303 too many blank lines (2)
./python-package/lightgbm/compat.py:140:5: E303 too many blank lines (2)
```

**Review by jameslamb — APPROVED** [review]
looks good, thanks for the help!

**Review by StrikerRUS — APPROVED** [review]

**Pull request merged**